### PR TITLE
rfc: update: bnorm: single-pass formula for calculating mean and variance (#1065)

### PR DESCRIPTION
**Pull request by igorsafo**
These changes extend the original RFC with information that was placed "here and there" in the discussions for the feature;

The changes don't change the initial proposal with adding an experimental build, but they clarify behavior for integration and validation of the feature:
- Extend verbose to emphasize an experimental build
- Modify benchdnn to avoid false-positive issues
- Enable all experimental features by default in an experimental build
- Extend regular testing to cover experimental builds
- Move experimental features to default build based on feedback from users

The motivation for a separate build:
- least impact to end users (we limit experimental build for frameworks use to experiment)
- zero-effort integration (once experimental build is obtained all experiment features are enabled by default)

[Rendered document](https://github.com/igorsafo/oneDNN/tree/rfcs/rfcs/20210519-single-pass-bnorm)

### CODE REVIEW

File: rfcs/20210519-single-pass-bnorm/README.md [hunk 1]
```diff
@@ -1,44 +1,81 @@
...
+
+The feature is intended to be used mostly by framework teams.
+
+### Experimental build
+To guard regular oneDNN users from accuracy issues but provide it to the
+frameworks for experiments a separate experimental build is proposed. No binary
+distribution is expected for experimental build. Users that
+want to try experimental features will have to build oneDNN on their side.
+**Once experimental build is obtained all experimental features are enabled**.
+To control features separately environmental variables are introduced for each
```
Comment (mgouicem) [inline rfcs/20210519-single-pass-bnorm/README.md:54]:
```suggestion
To control features separately environment variables are introduced for each
```

File: rfcs/20210519-single-pass-bnorm/README.md [hunk 2]
```diff
@@ -1,44 +1,81 @@
...
 configured to calculate those two values in inference as well.
 
+
+## Integration
+
+The feature is intended to be used mostly by framework teams.
+
+### Experimental build
+To guard regular oneDNN users from accuracy issues but provide it to the
+frameworks for experiments a separate experimental build is proposed. No binary
```
Comment (mgouicem) [inline rfcs/20210519-single-pass-bnorm/README.md:50]:
```suggestion
To provide this feature to framework teams for experimentation without impacting the accuracy of oneDNN for other users, we propose to use a separate experimental build. No binary
```

File: rfcs/20210519-single-pass-bnorm/README.md [hunk 3]
```diff
@@ -55,42 +92,41 @@ large in relation to result of the formula. In order to minimize rounding
...
-- Kahan summation algorithm will be used
+- Kahan summation algorithm [[#1]][1] will be used
 
 The last line, max( ) is there to make sure rounding errors won't cause
-variance to go negative. Further in batchnorm calculations square root is taken
+variance to go negative. Further in BN calculations square root is taken
 from variance. Negative value would result in NaN.
 
 New formula is considered experimental for now. By default the old one will be
 used. There needs to be a way for user to choose formula.
```
Comment (mgouicem) [inline rfcs/20210519-single-pass-bnorm/README.md:102]: shall this paragraph go away (you clarified this in the above integration section)

File: rfcs/20210519-single-pass-bnorm/README.md [hunk 4]
```diff
@@ -1,44 +1,81 @@
...
+frameworks for experiments a separate experimental build is proposed. No binary
+distribution is expected for experimental build. Users that
+want to try experimental features will have to build oneDNN on their side.
+**Once experimental build is obtained all experimental features are enabled**.
+To control features separately environmental variables are introduced for each
+experimental feature.
+
+- A new build flag `DNNL_EXPERIMENTAL` is introduced, **disabled by default**;
+- To opt-out BN single pass an environment variable
+`DNNL_EXPERIMENTAL_BNORM_STATS_ONE_PASS` is introduced;
```
Comment (mgouicem) [inline rfcs/20210519-single-pass-bnorm/README.md:59]:
```suggestion
`DNNL_EXPERIMENTAL_BNORM_STATS_ONE_PASS=<true*|false>` is introduced;
```

**Review by mgouicem — APPROVED** [review]

File: rfcs/20210519-single-pass-bnorm/README.md (hunk 1 above)
Comment (igorsafo) [reply rfcs/20210519-single-pass-bnorm/README.md]: Fixed, thanks!

File: rfcs/20210519-single-pass-bnorm/README.md (hunk 2 above)
Comment (igorsafo) [reply rfcs/20210519-single-pass-bnorm/README.md]: Rephrased, thanks

File: rfcs/20210519-single-pass-bnorm/README.md (hunk 4 above)
Comment (igorsafo) [reply rfcs/20210519-single-pass-bnorm/README.md:60]: fixed, thanks

File: rfcs/20210519-single-pass-bnorm/README.md (hunk 3 above)
Comment (igorsafo) [reply rfcs/20210519-single-pass-bnorm/README.md]: removed, thanks

**Pull request merged**